### ISSUE TITLE
Performances of some classes drop when training more epochs

### ISSUE THREAD

**Issue by chinchia** [nuclearboy95/Anomaly-Detection-PatchSVDD-PyTorch#10]
I don't quite understand why the performances of Carpet, Grid, Tile, and Screw classes getting worse and worse during training. The best performances of them almost all appear in the first epoch, and the Grid class even lasts at AUROC=0.5 for both detection and segmentation.
Does anyone knows why would this happen?

![image](https://user-images.githubusercontent.com/10960400/104138662-825f5e80-53e0-11eb-929c-4810593400bc.png)

**Comment by xiaxialin** [comment]
I have the same problem

**Comment by nuclearboy95** [comment]
For grid class, try lambda=1e-3 and lr=1e-4.

When the lambda is too large, the loss tends to squash the feature embeddings because of the Lsvdd term.
In this case, you will get a trivial solution (model outputs a constant).

When you get AUROC=0.500 for both detection and segmentation, it means that the model collapsed to a trivial solution.
So AUROC=0.500 is an indication of "lambda is too large".

As long as AUROC != 0.500, simply low AUROC does not mean a trivial solution because the model is outputting something meaningful.
For some classes (screw, in particular, checkout the fig below) the performance starts under 0.5 and goes up as training proceeds, so you may want to train for more epochs.
![image](https://user-images.githubusercontent.com/15180201/106579342-ee6a5800-6583-11eb-9032-7397c8b8cc11.png)

However, when lambda=0 (only self-supervised learning is used), AUROC stays below 0.5 forever.
So I think that the Lsvdd is important, but when lambda is too large, it *screws* everything up.

Btw, I used lambda=1.0 and lr=1e-4 for the screw class. Lambda=0.3 ~ 1.0 seems to be the sweet spot for the screw class.

**Comment by chinchia** [comment]
Hi author, thanks for your explanation. I tried lambda=1.0 for screw class and it did worked.

And I also tried lambda=1e-3 for grid class, but it still collapsed to a trivial solution and I get AUROC=0.5 for both detection and segmentation. I also tried lambda=1e-3 and lambda=1e-4 for both carpet and grid classes, but they all collapsed to trivial solutions. I don't quite understand. Will the AUROC increases if I run more than 300 epochs? Or do I need to adjust the lambda values?
